### ISSUE TITLE
[Feature] Add TensorBoard-Style Training Visualization

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#2663]
## Problem

Training metrics (loss, accuracy, learning rate) are only visible in console output. No persistent visualization for analysis.

### 1. Training Metrics Logger

```mojo
# shared/training/metrics_logger.mojo
struct MetricsLogger:
    var log_dir: String
    var metrics: Dict[String, List[Float32]]
    var step: Int
    
    fn __init__(out self, log_dir: String):
        self.log_dir = log_dir
        self.metrics = Dict[String, List[Float32]]()
        self.step = 0
        create_directory(log_dir)
    
    fn log_scalar(mut self, name: String, value: Float32):
        """Log a scalar metric."""
        if name not in self.metrics:
            self.metrics[name] = List[Float32]()
        
        self.metrics[name].append(value)
    
    fn log_dict(mut self, metrics: Dict[String, Float32]):
        """Log multiple metrics at once."""
        for name, value in metrics.items():
            self.log_scalar(name, value)
    
    fn step(mut self):
        """Increment step counter."""
        self.step += 1
    
    fn save(self):
        """Save all metrics to CSV files."""
        for name, values in self.metrics.items():
            var filename = self.log_dir + "/" + name + ".csv"
            write_csv(filename, values)
```

### 2. Usage in Training

```mojo
fn train():
    var logger = MetricsLogger("logs/lenet5_run1")
    
    for epoch in range(num_epochs):
        for batch in dataloader:
            var loss, acc = train_step(batch)
            
            logger.log_dict({
                "train/loss": loss,
                "train/accuracy": acc,
                "train/learning_rate": optimizer.lr
            })
            logger.step()
        
        # Validation
        var val_loss, val_acc = validate()
        logger.log_dict({
            "val/loss": val_loss,
            "val/accuracy": val_acc
        })
    
    logger.save()
```

### 3. CSV Format

```csv
# logs/lenet5_run1/train_loss.csv
step,value
0,2.301
1,2.145
2,1.987
...
```

### 4. Plotting Script

```python
# scripts/plot_training.py
"""Plot training metrics from CSV logs."""
import matplotlib.pyplot as plt
import pandas as pd
from pathlib import Path

def plot_training_run(log_dir: str):
    """Plot all metrics from a training run."""
    log_path = Path(log_dir)
    
    # Find all CSV files
    metrics = {}
    for csv_file in log_path.glob("*.csv"):
        name = csv_file.stem
        df = pd.read_csv(csv_file)
        metrics[name] = df
    
    # Create subplots
    fig, axes = plt.subplots(2, 2, figsize=(12, 8))
    
    # Plot train/val loss
    if 'train_loss' in metrics:
        axes[0, 0].plot(metrics['train_loss']['step'], 
                       metrics['train_loss']['value'], 
                       label='Train')
    if 'val_loss' in metrics:
        axes[0, 0].plot(metrics['val_loss']['step'], 
                       metrics['val_loss']['value'], 
                       label='Val')
    axes[0, 0].set_title('Loss')
    axes[0, 0].legend()
    
    # Plot train/val accuracy
    if 'train_accuracy' in metrics:
        axes[0, 1].plot(metrics['train_accuracy']['step'], 
                       metrics['train_accuracy']['value'], 
                       label='Train')
    if 'val_accuracy' in metrics:
        axes[0, 1].plot(metrics['val_accuracy']['step'], 
                       metrics['val_accuracy']['value'], 
                       label='Val')
    axes[0, 1].set_title('Accuracy')
    axes[0, 1].legend()
    
    # Plot learning rate
    if 'train_learning_rate' in metrics:
        axes[1, 0].plot(metrics['train_learning_rate']['step'], 
                       metrics['train_learning_rate']['value'])
    axes[1, 0].set_title('Learning Rate')
    
    plt.tight_layout()
    plt.savefig(log_path / 'training_plot.png')
    plt.show()

if __name__ == '__main__':
    import sys
    plot_training_run(sys.argv[1])
```

### 5. Usage

```bash
# Train model (generates CSV logs)
mojo run examples/lenet5/train.mojo

# Plot results
python scripts/plot_training.py logs/lenet5_run1
```

### 6. Future: TensorBoard Integration

When Mojo supports TensorBoard format:

```mojo
# Write TensorBoard events
struct TensorBoardLogger(MetricsLogger):
    fn log_scalar(mut self, name: String, value: Float32):
        # Write to TensorBoard event file
        write_tensorboard_scalar(self.log_dir, name, value, self.step)

# View with TensorBoard
# $ tensorboard --logdir=logs/
```

## Benefits

- Persistent training history
- Visual analysis of training dynamics
- Compare multiple runs
- Identify overfitting, learning rate issues
- Share results easily

## Example Analysis

Plots reveal:
- **Overfitting**: Val loss increases while train loss decreases
- **Learning rate too high**: Loss oscillates
- **Learning rate too low**: Loss plateaus early
- **Convergence**: Both losses stabilize

## Impact

- **Severity**: LOW
- **Analysis**: Better training insights
- **Debugging**: Visualize training issues

**Comment by mvillmow** [comment]
## Detailed Implementation Plan

Limit reached · resets 2pm (America/Los_Angeles)

**Comment by mvillmow** [comment]
## Dependencies (AI-Analyzed)

**Priority**: P1

### Depends On
- #2661
- #2638

### Blocks
- #2672
- #2673

### Analysis
TensorBoard-style visualization depends on logging framework and checkpoint system (for saving metrics). Blocks web dashboard and benchmark dataset work that need metrics persistence.

---
*Analyzed by Claude Code dependency analyzer*
